### ISSUE TITLE
Constant dictionaries with arrays only appending first key referenced.

### ISSUE THREAD

**Issue by NationalityNZ** [godotengine/godot#61919]
### Godot version

Godot 4.0 Alpha 9

### System information

Windows 10

### Issue description

May be related (but posting in case it's not, since it's subtly different): https://github.com/godotengine/godot/issues/61274

Constant dictionaries are only updating the first referenced key, at least in regards to arrays. I believe that constant dictionaries and array are only constant in regards to their size, so I believe this should be allowed.

var dictionaries are working fine.

```
const dictionary : Dictionary = {"ENTRY1":[],"ENTRY2":[]}

func _ready():
	dictionary["ENTRY1"].append("Hello")
	dictionary["ENTRY2"].append("World")
	print(dictionary) #{"ENTRY1":["Hello", "World"], "ENTRY2":[]}
```
Expected output is ```{"ENTRY1":["Hello"], "ENTRY2":["World"]}```

### Steps to reproduce

Use code above.

**Comment by AThousandShips** [comment]
I did some testing and I think I've identified part of the problem, will dig deeper and see

If you instead initialize the dictionary with unique arrays:
`{ "ENTRY1":["A"], "ENTRY2":["B"] }`
you get a different result, it appears that when accessing a constant dictionary by the subscript it fetches a constant value and that value is indexed by its original value, so in the original case it looks up the empty array, and regardless of key that returns the first entry (seemingly the first empty array as if you create two constant dictionaries the problem extends to that)

I think the problem might lie in the folding of dictionaries or somewhere else there, I'll take a deeper look later today and see

**Comment by AThousandShips** [comment]
In summary, as far as I read it:
When accessing a dictionary through the subscript operator the analyzer reduces this to a reduced value, and declares it constant, here:
https://github.com/godotengine/godot/blob/21c79d75153733b84533957228d0c267ef340a80/modules/gdscript/gdscript_analyzer.cpp#L3272-L3282

This is then emitted as a constant value in compilation, here:
https://github.com/godotengine/godot/blob/21c79d75153733b84533957228d0c267ef340a80/modules/gdscript/gdscript_compiler.cpp#L247-L249

In codegen this constant value is fetched from an array of constant values, or inserted if not present, and emitted as an address to that constant, in this case it looks up or adds an empty array

The problem seems to be caused by that when using the subscript for the second key the value that is reduced is again an empty array, meaning that it returns the constant address, which points to the first instance of fetching an empty array

In other words, we get, for a dictionary with keys "A" and "B" both holding empty arrays:
```
dictionary["A"].append("Hello") -> const([]).append("Hello")
dictionary["B"].append("World") -> const([]).append("World")
```
operating on the same empty array.

I'm not familiar enough with the inner workings of GDScript to propose a solution but I hope this can help anyone who would
